cardano-cli/src/Cardano/CLI/Shelley/Key.hs

```diff
@@ -0,0 +1,161 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- | Shelley CLI option data types and functions for cryptographic keys.
+module Cardano.CLI.Shelley.Key
+  ( SigningKeyDecodeError (..)
+  , readSigningKeyFile
+  , readSigningKeyFileAnyOf
+  ) where
+
+import           Cardano.Prelude
+
+import           Control.Monad.Trans.Except.Extra (handleIOExceptT)
+import qualified Data.Text as Text
+import qualified Data.Text.Encoding as Text
+import qualified Data.Text.IO as Text
+
+import           Cardano.Api.TextView (TextViewError (..))
+import           Cardano.Api.Typed
+
+import           Cardano.CLI.Types
+
+-- | Signing key decoding error.
+data SigningKeyDecodeError
+  = SigningKeyTextEnvelopeError !TextEnvelopeError
+  -- ^ The provided data seems to be a valid text envelope, but some error
+  -- occurred in extracting a valid signing key from it.
+  | SigningKeyBech32DecodeError !Bech32DecodeError
+  -- ^ The provided data is valid Bech32, but some error occurred in
+  -- deserializing it to a signing key.
+  | SigningKeyInvalidError
+  -- ^ The provided data does not represent a valid signing key.
+  deriving (Eq, Show)
+
+-- | Render an error message for a 'SigningKeyDecodeError'.
+renderSigningKeyDecodeError :: SigningKeyDecodeError -> Text
+renderSigningKeyDecodeError err =
+  case err of
+    SigningKeyTextEnvelopeError textEnvErr ->
+      Text.pack (displayError textEnvErr)
+    SigningKeyBech32DecodeError decodeErr ->
+      Text.pack (displayError decodeErr)
+    SigningKeyInvalidError -> "Invalid signing key."
+
+instance Error SigningKeyDecodeError where
+  displayError = Text.unpack . renderSigningKeyDecodeError
+
+-- | Read a signing key from a file.
+--
+-- The contents of the file can either be Bech32-encoded, hex-encoded, or in
+-- the text envelope format.
+readSigningKeyFile
+  :: forall keyrole.
+     ( HasTextEnvelope (SigningKey keyrole)
+     , SerialiseAsBech32 (SigningKey keyrole)
+     )
+  => AsType keyrole
+  -> SigningKeyFile
+  -> IO (Either (FileError SigningKeyDecodeError) (SigningKey keyrole))
+readSigningKeyFile asType (SigningKeyFile fp) =
```

Comment:
I wonder if it'd make more sense to attempt to detect the encoding of the file and then try deserializing it in the appropriate way.